pkg/controller/elasticsearch/reconcile/resources_state.go

```diff
@@ -25,6 +27,8 @@ type ResourcesState struct {
 	CurrentPodsByPhase map[corev1.PodPhase][]corev1.Pod
 	// DeletingPods are all deleted Elasticsearch pods.
 	DeletingPods []corev1.Pod
+	// StatefulSets are all existing StatefulSets for the cluster.
+	StatefulSets sset.StatefulSetList
```

Comment:
It may feel a bit strange to retrieve StatefulSets here for the only purpose of reporting the version in the status, but I think it's still way cleaner than propagating the k8s client to the function that updates the status data, that did not really need it.
I'm also thinking we should invest more in this struct, instead of retrieving the Pods and StatefulSets several times in a single reconciliation. See https://github.com/elastic/cloud-on-k8s/issues/3490.